File: leetcode/cpp/Trie.cpp

// leetcode.com 208. Implement Trie (Prefix Tree)

#include <string>
#include <vector>

using namespace std ;

class TrieNode {
public:
    TrieNode *child[26] ;
    
    bool isWord ;
    
    TrieNode() : isWord(false)
    {
        for(auto& s : child )  s = nullptr ;
    }
    
};

class Trie {
    
private:
    TrieNode *root ;
    
public:
    /** Initialize your data structure here. */
    Trie() {
        root = new TrieNode() ;
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        TrieNode *p = root ;
        
        for(auto &c : word)
        {
            int i = c - 'a' ;
            
            if(! p->child[i])  p->child[i] = new TrieNode() ;
            
            p = p->child[i] ;
            
        }
        
        p->isWord = true ;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        TrieNode *p = root ;
        
        for(auto& c : word)
        {
            int i = c - 'a' ;
            
            if( ! p->child[i])  return false ;
            
            p = p->child[i] ;
        }
        
        return p->isWord ;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        TrieNode *p = root ;
        
        for(auto& c : prefix)
        {
            int i = c - 'a' ;
            
            if( ! p->child[i])  return false ;
            
            p = p->child[i] ;
        }
        
        return true ;
        
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p200/Trie.java
package com.vkeonline.leetcode.p200;

/**
 * Leetcode [M]: 208. Implement Trie (Prefix Tree)
 * @author csgear
 */
public class Trie {
    private class Node {
        Node[] children = new Node[26]; // alphabet size
        boolean isEnd = false;
    }

    Node root;

    /**
     * Initialize your data structure here.
     */
    public Trie() {
        root = new Node();
    }

    /**
     * Inserts a word into the trie.
     */
    public void insert(String word) {
        Node current = root;
        for (char ch : word.toCharArray()) {
            if (current.children[ch - 97] == null)
            {
                current.children[ch - 97] = new Node();
            }
            current = current.children[ch - 97];
        }
        current.isEnd = true;
    }

    /**
     * Returns if the word is in the trie.
     */
    public boolean search(String word) {
        return isPrefixOrWord(word, false);
    }

    /**
     * Returns if there is any word in the trie that starts with the given prefix.
     */
    public boolean startsWith(String prefix) {
        return isPrefixOrWord(prefix, true);
    }

    private boolean isPrefixOrWord(String word, boolean isPrefix) {
        Node current = root;
        for (char ch : word.toCharArray()) {
            if (current.children[ch - 97] != null)
            {
                current = current.children[ch - 97];
            }
            else {
                return false;
            }
        }
        return isPrefix || current.isEnd;
    }
}
